packages/serum/src/market.ts

```diff
@@ -1310,31 +1311,29 @@ export class Market {
     const transaction = new Transaction();
     const signers: Account[] = [];
 
-    let wrappedSolAccount: Account | null = null;
+    let wrappedSolAccount: PublicKey | null = null;
     if (
       (this.baseMintAddress.equals(WRAPPED_SOL_MINT) &&
         baseWallet.equals(openOrders.owner)) ||
       (this.quoteMintAddress.equals(WRAPPED_SOL_MINT) &&
         quoteWallet.equals(openOrders.owner))
     ) {
-      wrappedSolAccount = new Account();
-      transaction.add(
-        SystemProgram.createAccount({
-          fromPubkey: openOrders.owner,
-          newAccountPubkey: wrappedSolAccount.publicKey,
-          lamports: await connection.getMinimumBalanceForRentExemption(165),
-          space: 165,
-          programId: TOKEN_PROGRAM_ID,
-        }),
+      wrappedSolAccount = await Token.getAssociatedTokenAddress(
+        ASSOCIATED_TOKEN_PROGRAM_ID,
+        TOKEN_PROGRAM_ID,
+        WRAPPED_SOL_MINT,
+        openOrders.owner,
       );
       transaction.add(
-        initializeAccount({
-          account: wrappedSolAccount.publicKey,
-          mint: WRAPPED_SOL_MINT,
-          owner: openOrders.owner,
-        }),
+        Token.createAssociatedTokenAccountInstruction(
```

Comment:
Will this tx now fail if the user already has a wSOL ATA?

Reply:
You know, that's an interesting question. I was mostly just replacing old code, but the old code also removes the account after completion [here](https://github.com/openbook-dex/openbook-ts/pull/2/files/cf6af9e18c283ebd2bba26e21886b74093840882#diff-aef00f531d41a0b6b7c6460464513b3f496659c1e083cbf4a13b6633d2219224L1362). 

That all being said, looking at the 2022 ATA code, it [looks like the operation is idempotent](https://github.com/solana-labs/solana-program-library/blob/c7c12f987c8794333f01f71604f7e742a89b0db5/associated-token-account/program/src/tools/account.rs#L18)

Reply:
Are you sure? Just tried on devnet to create the same account multiple times and hit this error https://github.com/solana-labs/solana-program-library/blob/master/associated-token-account/program/src/processor.rs#L109

I think CreateMode::Idempotent needs to be specified in the instruction data, otherwise it defaults to CreateMode::Always. The new createAssociatedTokenAccountIdempotentInstruction method (not sure if released yet) passes '1' in the instruction data https://github.com/solana-labs/solana-program-library/blob/master/token/js/src/instructions/associatedTokenAccount.ts#L48

Reply:
I think the only way to do this with idempotence with the current client library version is to use [getOrCreateAssociatedTokenAccount](https://solana-labs.github.io/solana-program-library/token/js/modules.html#getOrCreateAssociatedTokenAccount) (not atomic). Using an ATA and closing the account at the end of the transaction as before could fail if the ATA already exists.

Reply:
Tested locally and decided using the ATA doesn't make sense because of your above point (we close it anyway!) Not sure what was going on in the original, the change might have been a red herring to begin with.